README
NAME

Time::UTC::Now - determine current time in UTC correctly

DESCRIPTION

This module is one answer to the question "what time is it?".
It determines the current time on the UTC scale, handling leap seconds
correctly, and puts a bound on how inaccurate it could be.  It is the
rigorously correct approach to determining civil time.  It is designed
to interoperate with Time::UTC, which knows all about the UTC time scale.

UTC (Coordinated Universal Time) is a time scale derived from
International Atomic Time (TAI).  UTC divides time up into days, and
each day into seconds.  The seconds are atomically-realised SI seconds,
of uniform length.  Most UTC days are exactly 86400 seconds long,
but occasionally there is a day of length 86401 s or (theoretically)
86399 s.  These leap seconds are used to keep the UTC day approximately
synchronised with the non-uniform rotation of the Earth.  (Prior to 1972
a different mechanism was used for UTC, but that's not an issue here.)

Because UTC days have differing lengths, instants on the UTC scale
are identified here by the combination of a day number and a number
of seconds since midnight within the day.  In this module the day
number is the integral number of days since 1958-01-01, which is the
epoch of the TAI scale which underlies UTC.  This is the convention
used by the "Time::UTC" module.  That module has some functions to
format these numbers for display.  For a more general solution, use the
"utc_day_to_mjdn" function to translate to a standard Modified Julian
Day Number or the "utc_day_to_cjdn" function to translate to a standard
Chronological Julian Day Number, which can be used as input to a calendar
module.

INSTALLATION

	perl Build.PL
	./Build
	./Build test
	./Build install

AUTHOR

Andrew Main (Zefram) <zefram@fysh.org>

COPYRIGHT

Copyright (C) 2006, 2007, 2009, 2010
Andrew Main (Zefram) <zefram@fysh.org>

LICENSE

This module is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.


Changes
version 0.009; 2012-02-13

  * bugfix: require bugfixed version of Math::BigRat (for construction
    from decimal fraction strings)

  * bugfix: when loading Math::BigRat, do it in a way that isn't
    vulnerable to the bug in Perls prior to 5.12 regarding loading
    context-sensitive modules

  * bugfix: when calling Time::Unix::time(), don't be too picky about
    the internal flags on its return value

  * new now_utc_dec() function to return time in the decimal string form
    processed by Math::Decimal

  * refactor documentation for time determination functions

  * test that now_*() functions return well-formed results

  * test thread safety (except for thread cloning on Perl 5.6)

  * new internal architecture to better handle multiple time acquisition
    APIs

  * perform transformations of time data into output formats in XS code

  * new undocumented function to query all time acquisition APIs, for
    porting and debugging

  * in XS, declare "PROTOTYPES: DISABLE" to prevent automatic generation
    of unintended prototypes

  * correct dynamic_config setting to 0

  * include META.json in distribution

  * convert .cvsignore to .gitignore

  * add MYMETA.json to .cvsignore

version 0.008; 2010-10-20

  * in XS, use PERL_NO_GET_CONTEXT for efficiency

  * in XS, declare all local symbols static

  * load some modules at build time instead of configure time, to avoid
    dependency problems

  * check for required Perl version at runtime

  * use full stricture in test suite

  * in Build.PL, explicitly declare configure-time requirements

  * remove bogus "exit 0" from Build.PL

  * slightly reformat some C code to avoid exceeding 80 columns

  * add MYMETA.yml to .cvsignore

version 0.007; 2009-03-24

  * bugfix: require bugfixed versions of Data::Float and Module::Runtime
    (for $SIG{__DIE__} handling)

  * use Win32 API GetSystemTimeAsFileTime() where available

  * when dealing with the BSD nanosecond-resolution ntp_gettime(),
    process the uncertainty bound with nanosecond resolution

  * document the origin of each API that the module attempts to use

  * separate code for different APIs into different C functions

  * use typemap to simplify handling of Boolean argument

  * much improved probe system

  * don't use ppport.h, because it wasn't doing anything for perl v5.6+

  * use simpler "parent" pragma in place of "base"

  * build with Module::Build instead of ExtUtils::MakeMaker

  * rearrange source tree to fit Module::Build convention

  * test POD syntax and coverage, and make a small style change in
    documentation to satisfy the coverage test

  * complete dependency list

  * include signature in distribution

  * in documentation, separate "license" section from "copyright" section

version 0.006; 2007-02-05

  * add the utc_day_to_mjdn() function, duplicated from Time::UTC,
    for interoperability with other code that deals with UT days

  * when falling back on time(), add half a second onto its return value,
    because it rounds down

version 0.005; 2007-01-14

  * handle FreeBSD variations of ntp_gettime() interface: use of struct
    timespec instead of struct timeval; time may be returned with
    nanosecond precision instead of microsecond; leap state returned
    in time_state member of struct ntptimeval instead of return value
    of ntp_gettime()

  * in documentation, list OS-specific experiences (so far for Cygwin,
    FreeBSD, Linux, and Solaris)

  * when performing configuration tests in Makefile.PL, display the
    content of the test as well as the compilation command line

version 0.004; 2006-08-05

  * in now_utc_flt(), allow for the rounding error in converting the
    uncertainty value to floating point, as well as the rounding error
    in converting the seconds value to floating point

  * in now_utc_flt(), when allowing for the rounding error in converting
    to floating point, use the actual parameters of the floating point
    type (from Data::Float) instead of assuming IEEE 754 double

  * use "=> 0" instead of "=> undef" in unversioned dependencies in
    Makefile.PL

version 0.003; 2006-06-29

  * use ntp_gettime() in addition to ntp_adjtime() on systems where
    struct timex does not include the `time' member (the current time):
    this situation has been seen on FreeBSD 5.4 and Solaris 10, and
    prevented compilation of the previous version

  * zero out entire timex structure before ntp_adjtime(), not just the
    `modes' member, for the benefit of systems that check validity of
    unused control variables: this problem caused fallback to
    gettimeofday() on affected systems

  * include XSLoader in dependency list in Makefile.PL

  * various documentation clarifications

  * discuss overflow of day number for now_utc_sna()

  * add Time::TAI::Now to "see also" list

version 0.002; 2006-05-19

  * add the utc_day_to_cjdn() function, duplicated from Time::UTC, to
    avoid the need to use that module in order to process the output of
    this one

  * documentation tweaks for clarity

version 0.001; 2006-05-17

  * bugfix: Unix epoch was wrong in the code, making all returned values
    a day out

version 0.000; 2006-05-16

  * initial released version


Makefile.PL
use Config qw(%Config);
use ExtUtils::MakeMaker;

sub compile_and_link_ok($$) {
	my($hdrs, $code) = @_;
	my $incs = join("", map { "#include <$_>\n" } @$hdrs);
	open(TESTSRC, ">conftest.c") or die $!;
	print TESTSRC "#include \"EXTERN.h\"\n".
		"#include \"perl.h\"\n".
		"#include \"XSUB.h\"\n".
		"#include \"ppport.h\"\n".
		$incs."int main(void) { $code }\n";
	close(TESTSRC);
	my $cmd = "$Config{cc} $Config{ccflags} -I$Config{archlibexp}/CORE ".
		"$Config{ldflags} -o conftest$Config{exe_ext} conftest.c";
	print "testing: ", (map { "<$_> " } @$hdrs), $code, "\n", $cmd, "\n";
	return !system($cmd);
}

print "configuring...\n";
my $defs = "";
$defs .= " -DQUSE_NTP_ADJTIME"
	if compile_and_link_ok(["sys/timex.h"],
		"struct timex tx; int st; st = ntp_adjtime(&tx);");
$defs .= " -DQHAVE_STRUCT_TIMEX_TIME"
	if compile_and_link_ok(["sys/timex.h"],
		"struct timex tx; tx.time.tv_sec = 0; return tx.time.tv_sec;");
$defs .= " -DQHAVE_STRUCT_TIMEX_TIME_TV_NSEC"
	if compile_and_link_ok(["sys/timex.h"],
		"struct timex tx; tx.time.tv_nsec = 0; return tx.time.tv_nsec;");
$defs .= " -DQHAVE_STRUCT_TIMEX_TIME_STATE"
	if compile_and_link_ok(["sys/timex.h"],
		"struct timex tx; tx.time_state = 0; return tx.time_state;");
$defs .= " -DQHAVE_STRUCT_NTPTIMEVAL_TIME_TV_NSEC"
	if compile_and_link_ok(["sys/timex.h"],
		"struct ntptimeval ntv; ntv.time.tv_nsec = 0; return ntv.time.tv_nsec;");
$defs .= " -DQHAVE_STRUCT_NTPTIMEVAL_TIME_STATE"
	if compile_and_link_ok(["sys/timex.h"],
		"struct ntptimeval ntv; ntv.time_state = 0; return ntv.time_state;");
$defs .= " -DQUSE_GETTIMEOFDAY"
	if compile_and_link_ok(["sys/time.h"],
		"struct timeval tv; int st; st = gettimeofday(&tv, NULL);");

WriteMakefile(
	NAME => "Time::UTC::Now",
	VERSION_FROM => "lib/Time/UTC/Now.pm",
	PREREQ_PM => {
		Data::Float => "0.000",
		Math::BigRat => "0.02",
		Module::Runtime => "0.001",
		Time::Unix => "1.02",
		XSLoader => 0,
	},
	DEFINE => $defs,
	clean => { FILES => "conftest*" },
);


lib/Time/UTC/Now.pm
=head1 NAME

Time::UTC::Now - determine current time in UTC correctly

=head1 SYNOPSIS

	use Time::UTC::Now
		qw(now_utc_rat now_utc_sna now_utc_flt now_utc_dec);

	($day, $secs, $bound) = now_utc_rat;
	($day, $secs, $bound) = now_utc_rat(1);
	($day, $secs, $bound) = now_utc_sna;
	($day, $secs, $bound) = now_utc_sna(1);
	($day, $secs, $bound) = now_utc_flt;
	($day, $secs, $bound) = now_utc_flt(1);
	($day, $secs, $bound) = now_utc_dec;
	($day, $secs, $bound) = now_utc_dec(1);

	use Time::UTC::Now qw(utc_day_to_mjdn utc_day_to_cjdn);

	$mjdn = utc_day_to_mjdn($day);
	$cjdn = utc_day_to_cjdn($day);

=head1 DESCRIPTION

This module is one answer to the question "what time is it?".
It determines the current time on the UTC scale, handling leap seconds
correctly, and puts a bound on how inaccurate it could be.  It is the
rigorously correct approach to determining civil time.  It is designed to
interoperate with L<Time::UTC>, which knows all about the UTC time scale.

UTC (Coordinated Universal Time) is a time scale derived from
International Atomic Time (TAI).  UTC divides time up into days, and
each day into seconds.  The seconds are atomically-realised SI seconds,
of uniform length.  Most UTC days are exactly 86400 seconds long,
but occasionally there is a day of length 86401 s or (theoretically)
86399 s.  These leap seconds are used to keep the UTC day approximately
synchronised with the non-uniform rotation of the Earth.  (Prior to 1972
a different mechanism was used for UTC, but that's not an issue here.)

Because UTC days have differing lengths, instants on the UTC scale
are identified here by the combination of a day number and a number
of seconds since midnight within the day.  In this module the day
number is the integral number of days since 1958-01-01, which is the
epoch of the TAI scale which underlies UTC.  This is the convention
used by the C<Time::UTC> module.  That module has some functions to
format these numbers for display.  For a more general solution, use
the C<utc_day_to_mjdn> function to translate to a standard Modified
Julian Day Number or the C<utc_day_to_cjdn> function to translate to a
standard Chronological Julian Day Number, which can be used as input to
a calendar module.

=cut

package Time::UTC::Now;

{ use 5.006; }
use warnings;
use strict;

our $VERSION = "0.009";

use parent "Exporter";
our @EXPORT_OK = qw(
	now_utc_rat now_utc_sna now_utc_flt now_utc_dec
	utc_day_to_mjdn utc_day_to_cjdn
);

require XSLoader;
XSLoader::load("Time::UTC::Now", $VERSION);

=head1 FUNCTIONS

=head2 Time determination

Each of these functions determines the current UTC time and returns it.
They vary in the form in which the time is returned.  In each case, the
function returns a list of three values.  The first two values identify
a current UTC instant, in the form of a day number (number of days since
the TAI epoch) and a number of seconds since midnight within the day.
The third value is an inaccuracy bound, as a number of seconds, or
C<undef> if no accurate answer could be determined.

If an inaccuracy bound is returned then the function is claiming to have
answered correctly, to within the specified margin.  That is, some instant
during the execution of the function is within the specified margin of
the instant identified.  (This semantic differs from older current-time
interfaces that are content to return an instant that has already passed.)
The inaccuracy bound describes the actual time represented in the return
values, not some internal value that was rounded to generate the return
values.

The inaccuracy bound is measured in UTC seconds; that is, in SI seconds
on the Terran geoid as realised by atomic clocks.  This differs from SI
seconds at the computer's location, but the difference is only apparent
if the computer hardware is significantly time dilated with respect to
the geoid.

If C<undef> is returned instead of an inaccuracy bound then the function
could not find a trustable answer.  Either the clock available was not
properly synchronised or its accuracy could not be established.  Whatever
time could be found is returned, but the function makes no claim that it
is accurate.  It should be treated with suspicion.  In practice, clocks
of this nature are especially likely to misbehave around leap seconds.

Each function will C<die> if it can't find a plausible time at all.
If the I<DEMAND_ACCURACY> parameter is supplied and true then it will
also die if it could not find an accurate answer, instead of returning
with C<undef> for the inaccuracy bound.

=over

=item now_utc_rat([DEMAND_ACCURACY])

All three return values are in the form of C<Math::BigRat> objects.

This retains full resolution, is future-proof, and is easy to manipulate,
but beware that C<Math::BigRat> is currently rather slow.  If performance
is a problem then consider using one of the functions below that return
the results in other formats.

=item now_utc_sna([DEMAND_ACCURACY])

The day number is returned as a Perl integer.  The time since midnight
and the inaccuracy bound (if present) are each returned in the form of
a three-element array, giving a high-resolution fixed-point number of
seconds.  The first element is the integral number of whole seconds, the
second is an integral number of nanoseconds in the range [0, 1000000000),
and the third is an integral number of attoseconds in the same range.

This form of return value is fairly efficient.  It is convenient for
decimal output, but awkward to do arithmetic with.  Its resolution is
adequate for the foreseeable future, but could in principle be obsoleted
some day.

It is presumed that native integer formats will grow fast enough to always
represent the day number fully; if not, 31 bits will overflow late in
the sixth megayear of the Common Era.  (Average day length by then is
projected to be around 86520 s, posing more serious problems for UTC.)

=item now_utc_flt([DEMAND_ACCURACY])

All the results are returned as native Perl numbers.  The day number is
returned as a Perl integer, with the same caveat as for C<now_utc_sna>.
The other two items are floating point numbers.

This form of return value is very efficient and easy to manipulate.
However, its resolution is limited, rendering it obsolete in the near
future unless floating point number formats get bigger.

=item now_utc_dec([DEMAND_ACCURACY])

Each of the results is returned in the form of a string expressing a
number as a decimal fraction.  These strings are of the type processed
by L<Math::Decimal>, and are always returned in L<Math::Decimal>'s
canonical form.

This form of return value is fairly efficient and easy to manipulate.
It is convenient both for decimal output and (via implicit coercion to
floating point) for low-precision arithmetic.  L<Math::Decimal> can be
used for high-precision arithmetic.  Its resolution is unlimited.

=back

=head2 Day count conversion

=over

=item utc_day_to_mjdn(DAY)

This function takes a number of days since the TAI epoch and returns
the corresponding Modified Julian Day Number (a number of days since
1858-11-17 UT).  MJDN is a standard numbering for days in Universal Time.
There is no bound on the permissible day numbers.

=cut

use constant _TAI_EPOCH_MJDN => 36204;

sub utc_day_to_mjdn($) {
	my($day) = @_;
	return _TAI_EPOCH_MJDN + $day;
}

=item utc_day_to_cjdn(DAY)

This function takes a number of days since the TAI epoch and returns
the corresponding Chronological Julian Day Number (a number of days
since -4713-11-24).  CJDN is a standard day numbering that is useful as
an interchange format between implementations of different calendars.
There is no bound on the permissible day numbers.

=cut

use constant _TAI_EPOCH_CJDN => 2436205;

sub utc_day_to_cjdn($) {
	my($day) = @_;
	return _TAI_EPOCH_CJDN + $day;
}

=back

=head1 TECHNIQUES

There are several interfaces available to determine the time on a
computer, and most of them suck.  This module will attempt to use the
best interface available when it runs.  It knows about the following:

=over

=item ntp_adjtime(), ntp_gettime()

These interfaces were devised for Unix systems using the Mills timekeeping
model, which is intended for clocks that are synchronised via NTP
(the Network Time Protocol).  The timekeeping model is detailed in
L<ftp://ftp.udel.edu/pub/people/mills/memos/memo96b.ps>.

These interfaces gives some leap second indications, and an inaccuracy
bound on the time returned.  Both are faulty in their raw form, but they
are corrected by this module.  (Those interested in the gory details are
invited to read the source.)  Resolution 1 us, or on some systems 1 ns.

=item GetSystemTimeAsFileTime()

This is part of the Win32 API of Microsoft Windows.

Misbehaves around leap seconds, and does not give an inaccuracy bound.
Resolution of the interface is 100 ns.

=item gettimeofday()

This is a long-standing Unix interface, so named because it was the
interface to the "time-of-day clock".

Misbehaves around leap seconds, and does not give an inaccuracy bound.
Resolution 1 us.

=item Time::Unix::time()

This is derived from the original Unix C<time()> function, which was
also adopted by the C library standard and by Perl.  Various systems
have different epochs and resolutions for the C<time()> function, so
it is not usable by this module on its own.  The C<Time::Unix> module
corrects for the varying epochs across OSes.

Misbehaves around leap seconds, and does not give an inaccuracy bound.
Resolution 1 s.

=back

The author would welcome patches to this module to make use of
high-precision interfaces, along the lines of C<ntp_adjtime()>, on
non-Unix operating systems.

=head1 OS-SPECIFIC NOTES

The author would appreciate reports of experiences with this module
under OSes not listed in this section.

=head2 Cygwin

Uses gettimeofday(), which gives resolution 1 us but no uncertainty
bound and is discontinuous at leap seconds.  There is no interface that
supplies an uncertainty bound or correct leap second handling.

=head2 FreeBSD

Experimental code (new in version 0.005) uses the FreeBSD variation of
ntp_gettime(), which gives resolution 1 us or 1 ns (depending on system
configuration) and uncertainty bound.  Please report experiences with
this code to the author.

=head2 Linux

Uses ntp_adjtime(), which gives resolution 1 us and uncertainty bound.

=head2 Solaris

Uses ntp_gettime(), which gives resolution 1 us and uncertainty bound.

=head2 Windows

Experimental code (new in version 0.007) uses the native
GetSystemTimeAsFileTime().  Observed clock resolution is 10 ms, but
lower-order digits are supplied (filled with noise) down to the API
resolution of 100 ns.  There is no uncertainty bound, and there are
discontinuities at leap seconds.  There is no interface that supplies
an uncertainty bound or correct leap second handling.

=head1 SEE ALSO

L<Time::TAI::Now>,
L<Time::UTC>

=head1 AUTHOR

Andrew Main (Zefram) <zefram@fysh.org>

=head1 COPYRIGHT

Copyright (C) 2006, 2007, 2009, 2010, 2012
Andrew Main (Zefram) <zefram@fysh.org>

=head1 LICENSE

This module is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut

1;


Build.PL
{ use 5.006; }
use warnings;
use strict;

use Module::Build;

Module::Build->subclass(code => q{
	unless(__PACKAGE__->can("cbuilder")) {
		*cbuilder = sub { $_[0]->_cbuilder or die "no C support" };
	}
	sub compile_c {
		my($self, $file, %args) = @_;
		unless($args{no_feature_defs}) {
			my $feat_file = $self->localize_file_path(
					"lib/Time/UTC/features");
			my $probe_file = $self->localize_file_path(
					"lib/Time/UTC/features.probe");
			unless($self->up_to_date($probe_file, $feat_file)) {
				require IO::File;
				require Data::Dumper;
				my $probe = do($probe_file) or
					die $@ || "can't read $probe_file: $!";
				my $feat_defs = $probe->($self);
				$self->add_to_cleanup($feat_file);
				my $feat_fh = IO::File->new($feat_file, "w")
					or die "can't write $feat_file: $!";
				local $Data::Dumper::Terse = 1;
				$feat_fh->printflush(Data::Dumper::Dumper(
							$feat_defs))
					or die "can't write $feat_file: $!";
			}
			my $feat_defs = do($feat_file) or
				die $@ || "can't read $feat_file: $!";
			$args{defines} =
				{ %{$args{defines} || {}}, %$feat_defs };
			$obj_file = $self->cbuilder->object_file($file);
			unless($self->up_to_date($feat_file, $obj_file)) {
				unlink $obj_file;
			}
		}
		return $self->SUPER::compile_c($file, %args);
	}
})->new(
	module_name => "Time::UTC::Now",
	license => "perl",
	configure_requires => {
		"Module::Build" => 0,
		"perl" => "5.006",
		"strict" => 0,
		"warnings" => 0,
	},
	build_requires => {
		"Data::Dumper" => 0,
		"ExtUtils::CBuilder" => "0.15",
		"IO::File" => 0,
		"Module::Build" => 0,
		"Test::More" => 0,
		"perl" => "5.006",
		"strict" => 0,
		"warnings" => 0,
	},
	requires => {
		"Data::Float" => "0.008",
		"Exporter" => 0,
		"Math::BigRat" => "0.02",
		"Module::Runtime" => "0.005",
		"Time::Unix" => "1.02",
		"XSLoader" => 0,
		"constant" => 0,
		"integer" => 0,
		"parent" => 0,
		"perl" => "5.006",
		"strict" => 0,
		"warnings" => 0,
	},
	dynamic_config => 1,
	meta_add => { distribution_type => "module" },
	create_makefile_pl => "passthrough",
	sign => 1,
)->create_build_script;

1;
